solidity/test/OwnableMOfNSet.t.sol

```diff
@@ -0,0 +1,250 @@
+
+import "forge-std/Test.sol";
+import {OwnableMOfNSet} from "../contracts/libs/OwnableMOfNSet.sol";
+import {TypeCasts} from "../contracts/libs/TypeCasts.sol";
+
+contract OwnableMOfNSetTest is Test {
+    using TypeCasts for address;
+    event ValueAdded(
+        uint32 indexed domain,
+        address indexed value,
+        uint256 length
+    );
+    event ValueRemoved(
+        uint32 indexed domain,
+        address indexed value,
+        uint256 length
+    );
+    event ThresholdSet(uint32 indexed domain, uint8 threshold);
+    event CommitmentUpdated(uint32 indexed domain, bytes32 commitment);
+
+    OwnableMOfNSet set;
+
+    function setUp() public {
+        set = new OwnableMOfNSet();
+    }
+
+    function testAdd(uint32 domain, address value) public {
+        vm.assume(value != address(0x0));
+        vm.expectEmit(true, true, true, true, address(set));
+        emit ValueAdded(domain, value, 1);
+        vm.expectEmit(true, true, false, true, address(set));
+        bytes32 commitment = keccak256(abi.encodePacked(uint8(0), [value]));
+        emit CommitmentUpdated(domain, commitment);
+        set.add(domain, value);
+    }
+
+    function testAddTwice(uint32 domain, address value) public {
+        vm.assume(value != address(0x0));
+        set.add(domain, value);
+        vm.expectRevert(bytes("contained"));
+        set.add(domain, value);
+    }
+
+    function testAddZero(uint32 domain) public {
+        vm.expectRevert(bytes("zero address"));
+        set.add(domain, address(0));
+    }
+
+    function testAddNonowner(
+        uint32 domain,
+        address value,
+        address newOwner
+    ) public {
+        vm.assume(newOwner != address(0x0) && newOwner != set.owner());
+        set.transferOwnership(newOwner);
+        vm.expectRevert(bytes("Ownable: caller is not the owner"));
+        set.add(domain, value);
+    }
+
+    function testAddMany(uint8 numDomains, uint8 numValues) public {
+        vm.assume(numDomains < 32 && numValues < 32);
+        uint32 firstDomain = 100_000;
+        address firstValue = 0x3b2949fFFa5DC0bb41492AeBd12A89B286339858;
+        uint32[] memory domains = new uint32[](numDomains);
+        address[][] memory values = new address[][](numDomains);
+        for (uint32 i = 0; i < numDomains; i++) {
+            uint32 domain = firstDomain + i;
+            domains[i] = domain;
+            values[i] = new address[](numValues);
+            for (uint8 j = 0; j < numValues; j++) {
+                address value = address(uint160(firstValue) + j);
+                values[i][j] = value;
+                vm.expectEmit(true, true, true, true, address(set));
+                emit ValueAdded(domain, value, j + 1);
+            }
+            bytes32 commitment = keccak256(
+                abi.encodePacked(uint8(0), values[i])
+            );
+            vm.expectEmit(true, true, true, true, address(set));
+            emit CommitmentUpdated(domain, commitment);
+        }
```

Comment:
isn't this basically just duplicating the implementation?

Reply:
Yeah more or less. Have other suggestions for how to test?